[windows] Prepare toolchain build to run tests for LLDB

weliveindetail opened the pull request:
This PR adds handling for the `-Test lldb` option, which allows to enable the LLDB test suite on Windows builds of the swift toolchain, fixing https://github.com/swiftlang/llvm-project/issues/9141

While we still see a lot of test failures, it's time to give the CI a try. Once we get to a stable subset of tests that pass, we can iterate to enable the remaining ones and increase coverage.

The PR adds a local Windows-specific mechanism to mark bulks of tests as XFAIL or SKIPPED. These tests are listed in file `utils/windows-llvm-lit-test-overrides.txt`. The goal is to get it down to a low number and remove the mechanism and file eventually.

weliveindetail commented:
This is a preview. We need to land https://github.com/swiftlang/llvm-project/pull/9370 first.

weliveindetail commented on utils/build.ps1:
```diff
@@ -1343,7 +1343,49 @@ function Build-Compilers() {
+        $MakeToolPath = "C:/Program Files (x86)/GnuWin32/bin/make.exe"
+        if (-not (Test-Path $MakeToolPath)) {
+          & winget install GnuWin32.Make
+        }
```
@compnerd Is this an acceptable way/location to install `make`? I assume `winget` is available on the CI system?

weliveindetail commented on utils/build.ps1:
```diff
@@ -175,6 +175,12 @@ $msbuild = "$VSInstallRoot\MSBuild\Current\Bin\$BuildArchName\MSBuild.exe"
+# The make tool isn't part of MSYS, but we need it for LLDB tests
+$MakeToolPath = "C:/Program Files (x86)/GnuWin32/bin/make.exe"
+if (-not (Test-Path $MakeToolPath)) {
+  & winget install GnuWin32.Make
```
Alternatively, we could also use https://scoop.sh/ and run `scoop install main/make`. This is what the upstream bot https://lab.llvm.org/buildbot/#/builders/141 does and they dump:
```
-- Found make: C:/Users/tcwg/scoop/shims/make.exe
```

weliveindetail commented:
Oh wow, that works indeed! And sorry, it was easy to confuse. We need the *latest* linked PR. The other one was merged in the meantime :) But not too bad, I anyway had to patch `build-windows-toolchain.bat`.

@compnerd I checked locally and I think that the following will work:
```
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test windows platform
```

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test windows platform

weliveindetail commented:
Great, thanks. Good news is that `build.ps1` now runs with testing enabled for LLDB. Learning is that the build machine doesn't have `winget`, so we have to find another way to install `make.exe` from [GnuWin32]:(https://gnuwin32.sourceforge.net/packages/make.htm):
```
swift\utils\build.ps1 : The term 'winget' is not recognized as 
the name of a cmdlet, function, script file, or operable program.
```

The options I see:
* [Install winget](https://learn.microsoft.com/en-us/windows/package-manager/winget/#install-winget-on-windows-sandbox) and use it to install make.
* Check if we have [scoop](https://scoop.sh/#/apps?q=make&id=c43ff861c0f1713336e5304d85334a29ffb86317) and use it to install make.
* Check that the build machine provides make on Path and rely on it.
* Disable LLDB API tests that require make.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test windows platform

weliveindetail commented:
The build failed with the following error in build step `cmake.exe --build T:\5 --target distribution`:
```
[4797/6491][ 73%][1334.403s] Linking Swift static library lib\swiftASTGen.lib
...
swift\lib\ASTGen\Sources\ASTGen\SourceFile.swift:69:45: error: type 'Parser.ExperimentalFeatures' has no member 'sendingArgsAndResults'
    mapFeature(.SendingArgsAndResults, to: .sendingArgsAndResults)
                                           ~^~~~~~~~~~~~~~~~~~~~~
```
This doesn't appear to be related to my changes. Is it due to the merge conflict in `utils/build-windows-toolchain.bat` maybe? It's not related either, but does it cause the bot to silently skip the merge and build my (outdated) branch against newer dependencies? Let me resolve that.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test

weliveindetail commented:
Great, tests are running!
```
Testing Time: 202.24s
  Excluded           :   18
  Unsupported        : 1187
  Passed             : 2058
  Expectedly Failed  :   45
  Unresolved         :    1
  Failed             :   56
  Unexpectedly Passed:    4
```

SwiftREPL failures (28) will be fixed with https://github.com/swiftlang/llvm-project/pull/9421, the others I will investigate today.

weliveindetail commented:
In these NativePDB tests, a line breakpoint is set and it gets resolved to a location that seems right, but at startup time we add another location and that fails the test. This might be an actual bug:
```
  lldb-shell :: SymbolFile/NativePDB/local-variables.cpp
  lldb-shell :: SymbolFile/NativePDB/stack_unwinding01.cpp
```

The tests in https://github.com/swiftlang/swift/pull/76894/commits/25f5eca4263a5fcea8b8dfe9bd4de4ad00a148fe should be fixed and they pass reliably on my machine, but they failed in the CI test run. Let's see if they XPASS next time, because the companion PR in llvm-project now has the SDKROOT patch: https://github.com/swiftlang/llvm-project/pull/9429/commits/62b96261f6cac013ba6d87ec824fcbfd4b3ca6d6 Otherwise we might be missing a cherry-pick.

And for the record: The following test failed but the detailed report is lacking in the log! 🤷‍♂️ This is the only unreasonable difference to the log [from the companion PR](https://github.com/swiftlang/llvm-project/pull/9429#issuecomment-2436170027).
```
  lldb-api :: source-manager/TestSourceManager.py
```

weliveindetail commented:
The latest two commits integrate the changes from PRs https://github.com/swiftlang/swift/pull/76974 and https://github.com/swiftlang/swift/pull/77005 into this one. I would usually prefer small isolated changes, but the per-merge infrastructure makes it unbearable. I will abaondon the other PRs and only continue with this one.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test

weliveindetail commented:
SwiftREPL tests keep failing:
```
$ "t:\5\bin\lldb.exe" "--no-lldbinit" "-S" "T:/5/tools/lldb\test\Shell\lit-lldb-init-quiet" "--repl"
error: failed to stop process at REPL breakpoint
```
This means that the dummy executable for the REPL didn't load. It usually happens if it lacks a runtime dependency like swiftCore: https://github.com/swiftlang/llvm-project/blob/stable/20230725/lldb/tools/repl/swift/main.c#L66

weliveindetail commented:
Otherwise only `windows-unaligned-x86_64` failed this time:
```
llvm-project\lldb\test\Shell\Unwind\windows-unaligned-x86_64.test:20:10: error: CHECK: expected string not found in input
# CHECK: stop reason = breakpoint 1.1
         ^
<stdin>:9:61: note: scanning from here
Breakpoint 1: where = windows-unaligned-x86_64.test.tmp`func at windows-unaligned-x86_64.cpp:3, address = 0x0000000140001000
                                                            ^
<stdin>:13:14: note: possible intended match here
* thread #1, stop reason = breakpoint 1.2
             ^
```

This looks very similar to the problem we saw in the [NativePDB tests last time](https://github.com/swiftlang/swift/pull/76894#issuecomment-2437474806). `windows-unaligned-x86_64` would have failed for the same reason already, but it showed up as UNSUPPORTED back then. This indicates a flake at build-time of the inferior. It would be a pity if we had to skip it due to these issues, because there is a good chance that the actual bug behind it was fixed with https://github.com/llvm/llvm-project/issues/101710

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9429

@swift-ci please test

weliveindetail commented:
The builds failed with compile errors that indicate a LLVM version mismatch. I guess we have to build against `stable/20240723` now. I will prepare a new companion PR for that.

weliveindetail commented:
Can we run again with this PR please? This time we should reach the tests.
https://github.com/swiftlang/llvm-project/pull/9513

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test

compnerd commented on utils/build.ps1:
```diff
@@ -728,6 +728,12 @@ function Fetch-Dependencies {
+  $GnuWin32MakeURL = "https://downloads.sourceforge.net/project/ezwinports/make-4.4.1-without-guile-w32-bin.zip"
+  $GnuWin32MakeHash = "fb66a02b530f7466f6222ce53c0b602c5288e601547a034e4156a512dd895ee7"
+  DownloadAndVerify $GnuWin32MakeURL "$BinaryCache\GnuWin32Make-4.4.1.zip" $GnuWin32MakeHash
+  Extract-ZipFile GnuWin32Make-4.4.1.zip $BinaryCache GnuWin32Make-4.4.1
```
Can we limit this to just the case that `TestLLDB` is enabled?

compnerd commented on utils/build.ps1:
```diff
@@ -1435,7 +1448,54 @@ function Build-Compilers() {
+        # Install packages that the test suite requires
+        & $python -m pip install psutil
+        & $python -m pip install packaging
+        & $python -m pip install unittest2
```
Can we be more explicit about the user installation? (`$python -m pip install --user ...`).

compnerd commented on utils/build.ps1:
```diff
@@ -1435,7 +1448,54 @@ function Build-Compilers() {
+
+        $RuntimeBinaryCache = Get-TargetProjectBinaryCache $Arch Runtime
+
+        # Transitive dependency if _lldb.pyd
```
```suggestion
        # Transitive dependency of `_lldb.pyd`
```

compnerd commented on utils/build.ps1:
```diff
@@ -1435,7 +1448,54 @@ function Build-Compilers() {
+        cp $RuntimeBinaryCache\bin\swiftCore.dll "$CompilersBinaryCache\lib\site-packages\lldb"
+
+        # Runtime dependencies of repl_swift.exe
+        $SwiftrtSubdir = "lib\swift\windows"
```
```suggestion
        $SwiftRTSubdir = "lib\swift\windows"
```

weliveindetail commented:
@compnerd Thanks for your feedback.
@hjyamauchi Unfortunately, the bots didn't run. Can we try again please? We only need Windows platform right now.

```
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform
```

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

weliveindetail commented:
API tests turned UNRESOLVED again due to a Python issue. Modules are installed and discovered at configuration time, but when running `llvm-lit` they cannot be imported.

Again no new flaky tests. And `Unwind/windows-unaligned-x86_64.test` finally passed!

We have 4 remaining failures:
```
Failed Tests (4):
  lldb-unit :: Symbol/ClangArgs/UniquingCollisionWithAddedFlags
  lldb-unit :: Symbol/ClangArgs/UniquingCollisionWithExistingFlags
  lldb-unit :: Symbol/TestSwiftASTContext/ApplyWorkingDir
  lldb-unit :: Symbol/TestSwiftASTContext/PluginPath
```

We used to exclude them via `LIT_FILTER_OUT`, but it doesn't apply anymore. Maybe the syntax changed for unittests upstream? This is to be done.

weliveindetail commented:
Good new: the Python fix works! Not so good: not all new FAILs and XPASSes are ringing a bell. I will investigate tomorrow.
```
********************
Failed Tests (4):
  lldb-api :: lang/c/local_variables/TestLocalVariables.py
  lldb-api :: python_api/address_range/TestAddressRange.py
  lldb-shell :: Commands/command-process-launch-user-entry.test
  lldb-unit :: Host/PipeTest/WriteWithTimeout

********************
Unexpectedly Passed Tests (5):
  lldb-api :: api/command-return-object/TestSBCommandReturnObject.py
  lldb-api :: api/multiple-targets/TestMultipleTargets.py
  lldb-api :: commands/expression/nested/TestNestedExpressions.py
  lldb-api :: lang/cpp/class-template-non-type-parameter-pack/TestClassTemplateNonTypeParameterPack.py
  lldb-api :: lang/cpp/class-template-type-parameter-pack/TestClassTemplateTypeParameterPack.py
```

weliveindetail commented:
There is some explanation for each of them, except one: `Host/PipeTest/WriteWithTimeout` is mysterious. We can only skip the entire `HostTests.exe` unittests if we want to avoid it. But first, let's see if it fails again.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

weliveindetail commented:
No unexpected failures or passes. This should have been the first successful run:
```
Testing Time: 222.85s

Total Discovered Tests: 3421
  Excluded         :   20 (0.58%)
  Unsupported      : 1227 (35.87%)
  Passed           : 2055 (60.07%)
  Expectedly Failed:  116 (3.39%)
  Unresolved       :    3 (0.09%)
```
But still `Error: cmake.exe exited with code 1`. To be investigated next week.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

weliveindetail commented:
Finally, we have a successful run!
```
Total Discovered Tests: 3422
  Excluded         :   23 (0.67%)
  Unsupported      : 1228 (35.89%)
  Passed           : 2055 (60.05%)
  Expectedly Failed:  116 (3.39%)
```

Runtime impact on the CI pipeline was 5min: 3m10s for running the actual tests and ~2min for building LLDB test dependencies. Timing of the testing stage compared to the two surrounding successful builds:
```
    #33729: 19:25 - 19:57 = 32min (cxx interop)
--> #33732: 21:57 - 22:34 = 37min (enable lldb tests)
    #33733: 23:41 - 24:13 = 32min (enable libxml2)
```

weliveindetail committed:
[windows] Prepare toolchain build to run tests for LLDB

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test

weliveindetail commented:
Build failed for an unrelated reason. Didn't reach the test stage.

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test windows platform

compnerd commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test Windows platform

weliveindetail commented:
Tests finished successfully. Again no falkes, same results as last run, same timing:
```
#33837 23:00 - 23:37 = 37min
```

hjyamauchi commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test Windows platform

weliveindetail commented:
Rebase and fix conflicts after https://github.com/swiftlang/swift/pull/77948

compnerd commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test

weliveindetail commented:
The Windows build failed due to one test in LLDB:
```
======================================================================
FAIL: test_compressed_section_data (TestSectionAPI.SectionAPITestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llvm-project\lldb\packages\Python\lldbsuite\test\decorators.py", line 452, in wrapper
    return func(self, *args, **kwargs)
  File "C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llvm-project\lldb\test\API\python_api\section\TestSectionAPI.py", line 67, in test_compressed_section_data
    self.assertEqual(section_data, [0x20, 0x30, 0x40, 0x50, 0x60, 0x70, 0x80, 0x90])
AssertionError: Lists differ: [] != [32, 48, 64, 80, 96, 112, 128, 144]
```

Apparently, we got an empty list from `section.GetSectionData()`: https://github.com/swiftlang/llvm-project/blob/22b89865683de106191106cc504cda80db6d4dd1/lldb/test/API/python_api/section/TestSectionAPI.py#L66-L67

I never saw this test failing and I don't see a recent change, that might have caused it. Needs investigation. I will file an issue and disable it for the time being.

compnerd commented:
Please test with following PRs:
https://github.com/swiftlang/llvm-project/pull/9513

@swift-ci please test

weliveindetail commented:
LLDB tests finished successfully. No falkes. The build failed in the subsequent `check-swift` step, where the LLVM build-tree is reconfigured and LibXml2 can't be found.

weliveindetail commented:
@swift-ci please test

weliveindetail commented:
The Windows bot didn't run, let's try again.

@swift-ci please test Windows platform

weliveindetail commented:
Two flakes: first new, second known (we'd have to adjust the filter here, because it doesn't work for individual unittests)
```
Failed Tests (2):
  lldb-api :: commands/statistics/basic/TestStats.py
  lldb-unit :: Host/PipeTest/WriteWithTimeout
```

weliveindetail commented:
@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. Let's see if this is reproducible.

@swift-ci please test Windows platform

weliveindetail commented:
The build failed with an unrelated error in LLDB:
```
lldb\source\Plugins\LanguageRuntime\Swift\SwiftLanguageRuntimeDynamicTypeResolution.cpp(324): error C2039: 'MangleClass': is not a member of 'lldb_private::swift_demangle'
```

weliveindetail commented:
https://github.com/swiftlang/swift/issues/78488 should be fixed. Let's try again.

@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. Let's see if this is reproducible.

@swift-ci please test Windows platform

weliveindetail committed:
[dev] Enable tests for LLDB in CI and run them many times to detect flakes more efficiently

weliveindetail commented:
Tests passed. I added a temporary change to run `check-lldb` 10 times in each buildbot run to detect flakes more efficiently.

@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. Let's see if this is reproducible.

@swift-ci please test Windows platform

weliveindetail commented:
One test failed in first run of the test suite. It was added recently and passes upstream in bot lldb-aarch64-windows. It fails in swiftlang since https://github.com/swiftlang/llvm-project/pull/9493 relanded:
```
lldb-api :: functionalities/breakpoint/same_cu_name/TestFileBreakpoinsSameCUName.py
```

weliveindetail commented:
Added `TestFileBreakpoinsSameCUName.py` to filter and rebased.

@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. It should work the same way with the new order of test suites.

@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. Let's see if this is reproducible.

@swift-ci please test Windows platform

weliveindetail commented:
One failed test:
```
Failed Tests (1):
  lldb-api :: tools/lldb-server/TestNonStop.py
```

weliveindetail commented:
@swift-ci please test Windows platform

weliveindetail commented:
Apparently, tests failed but Swift CI is not reachable.

weliveindetail commented:
ci.swift.org is back online, but ci-external.swift.org still down

weliveindetail commented:
One test failed in first cycle. I am afraid we have to skip all lldb-server tests:
```
lldb-api :: tools/lldb-server/register-reading/TestGdbRemoteGPacket.py
```

weliveindetail commented:
@swift-ci please test Windows platform

weliveindetail commented:
The build failed because one test failed during setup:
```
Unresolved Tests (1):
  lldb-api :: lang/swift/async/tasks/TestSwiftTaskBacktrace.py
```

In particular, it failed due to `error: unknown argument: '-gdwarf'` in the following command for building the test subject:
```
T:/5/bin/swiftc \
  -emit-object C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llvm-project\lldb\test\API\lang\swift\async\tasks/main.swift     \
  -output-file-map T:\5\lldb-test-build.noindex\lang\swift\async\tasks\TestSwiftTaskBacktrace.test_dwarf/a-output-file-map.json      \
  -gdwarf -Onone -Xfrontend -serialize-debugging-options -parse-as-library -module-name a                                            \
  -module-cache-path T:/5/lldb-test-build.noindex/module-cache-clang\lldb-api -tools-directory "T:\5\bin"                            \
  -emit-module-path T:\5\lldb-test-build.noindex\lang\swift\async\tasks\TestSwiftTaskBacktrace.test_dwarf/a.swiftmodule              \
  -emit-module-interface-path T:\5\lldb-test-build.noindex\lang\swift\async\tasks\TestSwiftTaskBacktrace.test_dwarf/a.swiftinterface \
  -###

```

weliveindetail commented:
Tests passed. Let's see if this is reproducible.

@swift-ci please test Windows platform

weliveindetail commented:
Tests passed. Let's try a few more times, before landing this.

@swift-ci please test Windows platform

weliveindetail commented:
Polished, squashed and rebased changes. And I split off the CI batch file change that actually enables the tests. (The temporary `[dev]` commit is only here to keep exercising the CI.) I will prepare a second PR that enables the tests. In case we clog Swift PR testing, we can easily revert that one without loosing all the infrastructure.

weliveindetail commented:
@swift-ci please test Windows

weliveindetail committed:
Stop dumping unsupported tests. It's too verbose for production.

weliveindetail commented:
Build failure was unrelated: swift-driver configuration failed due to duplicate definition of imported target
```
CMake Error at T:/5/lib/cmake/llvm/LLVMExports.cmake:42 (message):
  Some (but not all) targets in this export set were already defined.

  Targets Defined: LLVMDemangle
```

weliveindetail commented:
@swift-ci please smoke test

weliveindetail commented:
A new test `lang/swift/async/tasks/TestSwiftTaskSelect.py` was added and it fails on Windows. Reported with https://github.com/swiftlang/llvm-project/issues/9887, adding a test override for it.

weliveindetail committed:
Skip test lang/swift/async/tasks/TestSwiftTaskSelect.py

weliveindetail commented:
New goal: land this on Tuesday morning

@swift-ci please test Windows

weliveindetail committed:
Revert "[dev] Enable tests for LLDB in CI and run them many times to detect flakes more efficiently"

This reverts commit 2ec68d311aed289d78f5c7120dca82ea4ac8280d.

weliveindetail commented:
Tests passed. I have to re-run the bots after reverting the temporary PR-testing changes. Let's get this PR landed and start testing tomorrow morning by merging https://github.com/swiftlang/swift/pull/78844 🤞

@swift-ci please smoke test

weliveindetail commented:
@swift-ci please smoke test Linux
@swift-ci please smoke test macOS

weliveindetail commented:
@swift-ci please smoke test Linux
